standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java

```diff
@@ -531,7 +531,7 @@ public void cleanTxnToWriteIdTable() throws MetaException {
         String s = "SELECT MIN(\"RES\".\"ID\") AS \"ID\" FROM (" +
             "SELECT MAX(\"TXN_ID\") + 1 AS \"ID\" FROM \"TXNS\" " +
             "UNION " +
-            "SELECT MIN(\"WS_COMMIT_ID\") AS \"ID\" FROM \"WRITE_SET\" " +
+            "SELECT MIN(\"WS_TXNID\") AS \"ID\" FROM \"WRITE_SET\" " +
```

Comment:
@zchovan  are you sure that is an equivalent change? @lcspinter what do you think?

Reply:
If my memory serves me well, `WS_COMMIT_ID` is the `NEXT_TXN_ID` at the time of the commit, and the `WS_TXNID` is the id of the given transaction

Reply:
that is correct. However, I don't understand how this is supposed to replace MIN_HISTORY table. Original code:
````
long minUncommittedTxnId = findMinOpenTxnGLB(stmt);
// If there are aborted txns, then the minimum aborted txnid could be the min_uncommitted_txnid
// if lesser than both NEXT_TXN_ID.ntxn_next and min(MIN_HISTORY_LEVEL .mhl_min_open_txnid).
String s = "SELECT MIN(\"TXN_ID\") FROM \"TXNS\" WHERE \"TXN_STATE\" = " + quoteChar(TXN_ABORTED);
minUncommittedTxnId = Math.min(minAbortedTxnId, minUncommittedTxnId);
```

Reply:
This was the original code, so we might want to check `findMinOpenTxnGLB`:
```
        // First need to find the min_uncommitted_txnid which is currently seen by any open transactions.
        // If there are no txns which are currently open or aborted in the system, then current value of
        // NEXT_TXN_ID.ntxn_next could be min_uncommitted_txnid.
        long minUncommittedTxnId = findMinOpenTxnGLB(stmt);

        // If there are aborted txns, then the minimum aborted txnid could be the min_uncommitted_txnid
        // if lesser than both NEXT_TXN_ID.ntxn_next and min(MIN_HISTORY_LEVEL .mhl_min_open_txnid).
        String s = "SELECT MIN(\"TXN_ID\") FROM \"TXNS\" WHERE \"TXN_STATE\" = " + TxnStatus.ABORTED;
        LOG.debug("Going to execute query <" + s + ">");
        rs = stmt.executeQuery(s);
        if (rs.next()) {
          long minAbortedTxnId = rs.getLong(1);
          if (minAbortedTxnId > 0) {
            minUncommittedTxnId = Math.min(minAbortedTxnId, minUncommittedTxnId);
          }
        }
```

Reply:
This was the `findMinOpenTxnGLB`
```
  /**
   * See doc at {@link TxnStore#findMinOpenTxnId()}
   * Note that {@link #openTxns(OpenTxnRequest)} makes update of NEXT_TXN and MIN_HISTORY_LEVEL
   * a single atomic operation (and no one else should update these tables except the cleaner
   * which deletes rows from MIN_HISTORY_LEVEL which can only allow minOpenTxn to move higher)
   */
  private long findMinOpenTxnGLB(Statement stmt) throws MetaException, SQLException {
    String s = "SELECT \"NTXN_NEXT\" FROM \"NEXT_TXN_ID\"";
    LOG.debug("Going to execute query <" + s + ">");
    ResultSet rs = stmt.executeQuery(s);
    if (!rs.next()) {
      throw new MetaException("Transaction tables not properly " +
          "initialized, no record found in next_txn_id");
    }
    long hwm = rs.getLong(1);
    s = "SELECT MIN(\"MHL_MIN_OPEN_TXNID\") FROM \"MIN_HISTORY_LEVEL\"";
    LOG.debug("Going to execute query <" + s + ">");
    rs = stmt.executeQuery(s);
    rs.next();
    long minOpenTxnId = rs.getLong(1);
    if(rs.wasNull()) {
      return hwm;
    }
    //since generating new txnid uses select for update on single row in NEXT_TXN_ID
    assert hwm >= minOpenTxnId : "(hwm, minOpenTxnId)=(" + hwm + "," + minOpenTxnId + ")";
    return minOpenTxnId;
  }
```

Reply:
So it think the theory is this:
- Original: Minimum open transactionId which was open when any of the currently executing transactions started.
- New: Minimum open transactionId which was open when the last write committed.

So originally we removed lines where the transaction older of any open/aborted transactions, and even the running transactions were not needed to see the older changes, but this was based on the starting time of the transaction.

With the new change we are only keeping the `WRITE_SET` based on the transactionId when the actual write was committed (not when the actual write was started), as the previous queries should not read the non-committed folder/data anyway

Question: What happens with the `WRITE_SET` when a compaction happened on this table (we are relying here on the fact that we are the only ones removing lines from this table)

Reply:
````With the new change we are only keeping the WRITE_SET based on the transactionId when the actual write was committed (not when the actual write was started), as the previous queries should not read the non-committed folder/data anyway````
that leads to the violation of the SNAPSHOT isolation, see `testUpdateSnapshotIsolation test

Reply:
````What happens with the WRITE_SET```` - nothing, it's only cleaned by the AcidHouseKeeperThread:
````
performTask(txnHandler::performWriteSetGC, "Cleaning obsolete write set entries");
performTask(txnHandler::cleanTxnToWriteIdTable, "Cleaning obsolete TXN_TO_WRITE_ID entries");
````

Reply:
+1